medusa/helpers.py

```diff
@@ -1189,7 +1189,19 @@ def get_url(url, post_data=None, params=None, headers=None, timeout=30, session=
     hooks, cookies, verify, proxies = request_defaults(kwargs)
     method = u'POST' if post_data else u'GET'
 
+    # Get RequestPolice instance object.
+    rpolice = kwargs.pop(u'rpolice', None)
+
+    from helper.request_police import RequestPoliceException
+
     try:
+        if rpolice:
+            try:
+                [rpolice_check() for rpolice_check in rpolice.enabled_police_request_hooks]
+            except RequestPoliceException as e:
```

Comment:
What I mean is instead of creating a bunch of hooks, create a custom session class from requests.Session like:
```
class PolicedSession(requests.Session):
```
And subclass the methods there.  That way they are all a part of the session directly.  Then you can use the PolicedSession in place of the requests.Session.  You can even hook the session.auth to get the pre-requisite API limits etc from the service.